pkg/ssl/sign.go

```diff
@@ -0,0 +1,260 @@
+/*
+Package ssl implements the Secure Systems Lab signing-spec (sometimes
+abbreviated SSL Siging spec.
+https://github.com/secure-systems-lab/signing-spec
+*/
+package ssl
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+)
+
+// ErrUnknownKey indicates that the implementation does not recognize the
+// key.
+var ErrUnknownKey = fmt.Errorf("unknown key")
+
+// ErrNoSignature indicates that an envelope did not contain any signatures.
+var ErrNoSignature = fmt.Errorf("no signature found")
+
+/*
+Envelope captures an envelope as described by the Secure Systems Lab
+Signing Specification. See here:
+https://github.com/secure-systems-lab/signing-spec/blob/master/envelope.md
+*/
+type Envelope struct {
+	PayloadType string      `json:"payloadType"`
+	Payload     string      `json:"payload"`
+	Signatures  []Signature `json:"signatures"`
+}
+
+/*
+Signature represents a generic in-toto signature that contains the identifier
+of the key which was used to create the signature.
+The used signature scheme has to be agreed upon by the signer and verifer
+out of band.
+The signature is a base64 encoding of the raw bytes from the signature
+algorithm.
+*/
+type Signature struct {
+	KeyID string `json:"keyid"`
+	Sig   string `json:"sig"`
+}
+
+/*
+Pae implementes PASETO Pre-Authentic Encoding
+https://github.com/paragonie/paseto/blob/master/docs/01-Protocol-Versions/Common.md#authentication-padding
+*/
+func Pae(data [][]byte) ([]byte, error) {
```

Comment:
Would this ever have more than two elements (Type and Data)? Would it be cleaner to do:

`Pae(payloadType string, payload []byte)`?

Reply:
In the case of SSL or ITE-5, no. But the PAE spec allows for arbitrary number of arguments, so I was honouring that here.
The function is called from from SSL envelope signer (and verifier) with the signature `func (es *EnvelopeSigner) SignPayload(payloadType string, body []byte) (*Envelope, error)` which is very similar to what you want. So from a client to SSL signing spec's view, the interface is `(payloadType string, payload []byte)`. 
And this is even simplified even further from in-toto's view (ITE-5 which mandates a constant payload type), so from a client to in-toto the interface is `func (s *SslSigner) SignPayload(body []byte) (*ssl.Envelope, error)`.

Reply:
Ah, makes sense. Thanks!